Add dos and monitoring docs (#160)
libp2p/docs: content/reference/dos-mitigation.md

---

> title: "DOS Mitigation"

Reviewer: Can we have a table of contents?  Or does the docs website autogenerate?  I didn't see one in https://bafybeiegvuphbhruoy6ntjhgh5xfmweye7ubeshinyv5lugdhgxml6dice.on.fleek.co/reference/dos-mitigation/

Author: added one

weight: 4
---

In general we have to design our protocols to be resilient to malicious or

> misbehaving nodes. libp2p offers reasonable defaults and primitives tools let the

Reviewer: ```suggestion
misbehaving nodes. libp2p offers reasonable defaults and primitives to let the
```

end application developer design resilient peer-to-peer programs.

> # General tactics that work across libp2p implementations

Reviewer: Looks like we need to add a level of indentation throughout given the rendering in https://bafybeiegvuphbhruoy6ntjhgh5xfmweye7ubeshinyv5lugdhgxml6dice.on.fleek.co/reference/dos-mitigation/

> ## Limit the number of concurrent streams your protocol needs

Reviewer 2: Optional: I think a similar section on the number of connections would be helpful.

Reviewer: Agreed.

> ## Limit the number of concurrent streams your protocol needs

Reviewer: To be clear that we're not talking about global number of streams for a protocol, but the number of streams per connection...

```suggestion
## Limit the number of concurrent streams per connection your protocol needs
```

Each stream has some resource cost associated with it. Depending on the
transport and multiplexer, this can be bigger or smaller. Try to avoid having
too many concurrent streams open per peer for your protocol. Instead try to
limit the maximum number of concurrent streams to something reasonable (surely
you don't need >512 streams open at once for a peer?). Multiple concurrent
streams can be useful for logic or to avoid [Head-of-line
blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking).

Using a stream for a short period of time and then closing it is fine. It's

> really the number of _concurrent_ streams that you need to be careful of.

Reviewer: Maybe use Identify as an example, where it doesn't make sense for the protocol to support more than one stream per connection?

## Reduce blast radius

If you can split up your libp2p application into multiple separate processes you
can increase the resiliency of your overall system. For example your node may
have to help achieve consensus and respond to user queries. By splitting this up
into two processes you now rely on the OS’s guarantee that the user query
process won’t take down the consensus process.

## Fail2ban

> If you can log when a peer is misbehaving or malicious, you can then hook up

Reviewer 2: ```suggestion
If you can log when a peer is misbehaving or is malicious, you can then hook up
```

those logs to fail2ban and have fail2ban manage your firewall to automatically
block misbehaving nodes. Go libp2p includes some builtin support for this
usecase. More details below.

> # Tactics for Go libp2p

Reviewer: ```suggestion
# Tactics for go-libp2p
```

This section is meant for an operator of a go-libp2p application. In the future
we hope that this advice will work equally well amongst all implementations.

> ## Resource Manager

Reviewer: This section seems incomplete.  At the minimum, we should be linking to the resource manager guide/docs.

go-libp2p includes a powerful resource manager that keeps track of every active
connection and stream. It exposes some useful metrics that can help you tune
your resource limits. It’s important to keep an eye on the number of blocked

> resource requests, if you are constantly blocking resources during normal

Reviewer: ```suggestion
resource requests.  If you are constantly blocking resources during normal
```

operation your limits are too low. But a very rare spike is a good sign that the
resource manager protected your node from something. You can hook these metrics
up to Prometheus (or any other opencensus exporter) and Grafana (we even include
a prebuilt dashboard). See

> [https://github.com/libp2p/go-libp2p-resource-manager/tree/master/obs/grafana-dashboards](https://github.com/libp2p/go-libp2p-resource-manager/tree/master/obs/grafana-dashboards)

Reviewer: ```suggestion
https://github.com/libp2p/go-libp2p-resource-manager/tree/master/obs/grafana-dashboards
```

> [https://github.com/libp2p/go-libp2p-resource-manager/tree/master/obs/grafana-dashboards](https://github.com/libp2p/go-libp2p-resource-manager/tree/master/obs/grafana-dashboards)

Reviewer: It seems like more of this should be in monitoring.md and we link there.

for more details on how to set this up.

![blocked-resource-screenshot](./assets/blocked-resource-screenshot.png)

## Rate limit incoming connections

Depending on your use case, it can help to limit the number of inbound

> connections. You can use go-libp2p's

Reviewer: Do we give guidance on when to use this mechanism vs. the resource manager?  I see this is a good hook for custom logic, but it seems like what Prysm is doing could be covered by go-libp2p resource manager right?

Author: > Do we give guidance on when to use this mechanism vs. the resource manager?

No, but I could add something here.

> I see this is a good hook for custom logic, but it seems like what Prysm is doing could be covered by go-libp2p resource manager right?

Not really. If you're trying to avoid an adversary that can connect to you and give you a ton of work to do all at once the rcmgr doesn't protect at all. This attack can easily be mitigated by rate limiting though.

Not all applications will want this rate limiting, or they may want to rate limit certain things (e.g. something in the protocol rather than in the connections). For example, if I'm Google I wouldn't want to rate limit any new connection to me. I would rather rate limit work per connection.

Should the rcmgr do this? I don't think so. It's not directly related to limiting the resources being used, and if it can be handled by a smaller component that already exists the better.

I hope that makes sense, but happy to expand more as well.

and `InterceptAccept` for this. For a concrete example, take a look at how Prysm
implements their (Connection
Gater)[https://github.com/prysmaticlabs/prysm/blob/63a8690140c00ba6e3e4054cac3f38a5107b7fb2/beacon-chain/p2p/connection_gater.go#L43].

## Who’s misbehaving?

To answer the question of which peer is misbehaving and harming you, go-libp2p
exposes a [canonical log
lines](https://github.com/libp2p/go-libp2p-core/blob/master/canonicallog/canonicallog.go#L18)
that identifies a misbehaving peers. A canonical log line is simply a log line
with a special format. For example here’s a peer status log line that tells us a
peer established a connection with us, and that this log line was randomly
sampled (1 out of 100).

```
Jul 27 12:14:14 ipfsNode ipfs[46133]: 2022-07-27T12:14:14.674Z        INFO        canonical-log        swarm/swarm_listen.go:128        CANONICAL_PEER_STATUS: peer=12D3KooWSbNLGMYeUuMSXDiHwbhXHzTJaWZzH95MZzeAob9BeB51 addr=/ip4/147.75.74.239/udp/4001/quic sample_rate=100 connection_status="established" dir="inbound"
```

To see these kinds of logs make sure you’ve enabled the `"canonical-log=info"`

> log level. You can this in code like

Reviewer 2: ```suggestion
log level. You can do this in code like
```

[so](https://github.com/libp2p/go-libp2p-core/blob/master/canonicallog/canonicallog_test.go#L14),
or by setting the environment variable `GOLOG_LOG_LEVEL="canonical-log=info"`.

## How to block a misbehaving peer

Once you’ve identified the misbehaving peer, you can block them with `iptables`
or `ufw`. Here we’ll outline how to block the peer with `ufw`. You can get the
ip address of the peer from the
[multiaddr](https://github.com/multiformats/multiaddr) in the log.

```bash
sudo ufw deny from 1.2.3.4
```

## How to automate blocking with fail2ban

> You can hook up [fail2ban](https://www.fail2ban.org/wiki/index.php/Main_Page) to

Reviewer: ```suggestion
You can hook up [fail2ban](https://www.fail2ban.org) to
```

automatically block connections from these misbehaving peers if they emit this
log line multiple times in some period of time. For example, a simple fail2ban
filter for go-libp2p would look like this:

```
[Definition]

> failregex = ^.*[\t\s]CANONICAL_PEER_STATUS: .* addr=\/ip[46]\/<HOST>[^\s]*

Reviewer: I had a question about this in https://www.notion.so/pl-strflt/Guide-for-how-to-respond-to-resource-exhaustion-attacks-b10f55cc9a3d4917ae80c9b914e05e8c.

Author: Answered. Copied here:

> `<HOST>`: I assume this is fail2ban syntax for saying that the host to block is this string that comes after “/ip[46]/” and before “[^\s]”.  Is that right?  (I was a little surprised not to see more conventional capture group conventions.)

Exactly. fail2ban expands this to a regex that captures the host `(?:::f{4,6}:)?(?P<host>\S+)`. See https://www.fail2ban.org/wiki/index.php/MANUAL_0_8#Filters

```
`/etc/fail2ban/filter.d/go-libp2p-peer-status.conf`

This matches any canonical peer status logs. If a peer shows up often in these
sampled logs, something abnormal is happening. i.e. maybe they are churning
connections.

A conservative fail2ban rule for go-libp2p using the above filter would look
like this:

```

> [go-libp2p-weird-behavior-iptables]

Reviewer: I had some comments about this section and format/layout changes in I had a question about this in https://www.notion.so/pl-strflt/Guide-for-how-to-respond-to-resource-exhaustion-attacks-b10f55cc9a3d4917ae80c9b914e05e8c.

Author: Answered 

> @Marco Munizaga : it’s not clear to me how the fail2ban rule below is tied to the filter above.  I don’t see an id or name reference.

By the filename. The filter is in a file go-libp2p-peer-status.conf and this rule references filter=go-libp2p-peer-status.

I'll leave a comment on the relevant line as well

[Unit]
After=network.target
Description=ipfs-daemon

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/r4jm7wfirgdr84zmsnq5qy7hvv14c7l7-glibc-locales-2.34-210/lib/locale/locale-archive"
Environment="PATH=/nix/store/7jr7pr4c6yb85xpzay5xafs5zlcadkhz-coreutils-9.0/bin:/nix/store/140f6s4nwiawrr3xyxarmcv2mk62m62y-findutils-4.9.0/bin:/nix/store/qd9jxc0q00cr7fp30y6jbbww20gj33lg-gnugrep-3.7/bin:/nix/store/lgvd2fh4cndlv8mnyy49jp1nplpml3xp-gnused-4.8/bin:/nix/store/0f3ncs289m2x1vmv2b3grd6l9x1yp2m3-systemd-250.4/bin:/nix/store/7jr7pr4c6yb85xpzay5xafs5zlcadkhz-coreutils-9.0/sbin:/nix/store/140f6s4nwiawrr3xyxarmcv2mk62m62y-findutils-4.9.0/sbin:/nix/store/qd9jxc0q00cr7fp30y6jbbww20gj33lg-gnugrep-3.7/sbin:/nix/store/lgvd2fh4cndlv8mnyy49jp1nplpml3xp-gnused-4.8/sbin:/nix/store/0f3ncs289m2x1vmv2b3grd6l9x1yp2m3-systemd-250.4/sbin"
Environment="TZDIR=/nix/store/n83qx7m848kg51lcjchwbkmlgdaxfckf-tzdata-2022a/share/zoneinfo"

Environment=GOLOG_LOG_LEVEL="canonical-log=info" LIBP2P_RCMGR=1
ExecStart=/nix/store/mmvd2akskpaszlradl8qv4v703v1cy11-kubo-0.0.1/bin/ipfs daemon
Restart=always
RestartSec=1min
User=ipfs
```

### Example screen recording of fail2ban in action

> [fail2ban+go-libp2p.mov](./assets/fail2bango-libp2p.mov)

Reviewer: Is .mov supported across platforms without quicktime?  Maybe do .mp4 instead?

> ## Setting Up fail2ban

Reviewer: I made some edits in I had a question about this in https://www.notion.so/pl-strflt/Guide-for-how-to-respond-to-resource-exhaustion-attacks-b10f55cc9a3d4917ae80c9b914e05e8c. about doing this more as a numbered list.

Author: Thanks! I copied those changes over here

> ## Setting Up fail2ban

Reviewer: ```suggestion
### Setting Up fail2ban
```

   isolated from the honest peers in the network. This is commonly called an
   eclipse attack and is viable if it's either cheap to eclipse this node, or if
   eclipsing this node has a high payoff.

Generally the effect on our application can range from crashing to stalling to
failing to handle new peers to degraded performance. Ideally we want
our application to at worst suffer a slight perfomance penalty, but otherwise
stay up and healthy.

In the next section we'll cover some design strategies you should incorporate
into your protocol to make sure your application stays up and healthy.

# Incoporating DOS mitigation from the start

The general strategy is to use the minimum amount of resources as possible, and
make sure that there's no untrusted amplification mechanism (e.g. an untrusted
node can force you do to 10x the work it does). A protocol level reputation
system can help (take a look at [GossipSub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) for inspiration) as well as
logging misbehaving nodes and actioning those logs separately (see fail2ban
below).

Here are some more specific recommendations

> ## Limit the number of concurrent streams per connection your protocol needs

Reviewer: Can we give pointers on how to do this? 

For go-libp2p this means using resource manager right?

For rust, this isn't at the connection level, but somewhere I think we should be linking to https://docs.rs/libp2p/latest/libp2p/swarm/struct.SwarmBuilder.html#method.max_negotiating_inbound_streams .  Maybe we say, "rust-libp2p relies on each protocol to limit the number of streams per connection in XXX.  A global upperbound on negotiating/transient inbound streams can be set using https://docs.rs/libp2p/latest/libp2p/swarm/struct.SwarmBuilder.html#method.max_negotiating_inbound_streams."

Author: I should be more clear.

Here I'm talking about limiting the number of concurrent streams you need by design of the protocol, as opposed to using an existing protocol and trying to limit the streams at the end. For example imagine a RPC style protocol whose procedures are async and often take a long time to return (say > 1min). Here are two ways you could implement it:

1. Open a stream for each RPC call, and keep that stream open until the rpc call returns.
2. Open a stream for the start of the call then close it. The remote side will open a new stream with the answer.

Assume you make a lot of concurrent calls, method 1 would result in a large number of concurrent and mostly inactive streams. Method 2 would result in a fewer number of concurrent streams, and thus lower memory footprint.

If you add a limit here of say 10 streams, then method 1 will mean you can only have 10 concurrent RPC calls, while method 2 would let you have a much larger number of concurrent RPC calls.

Does that make sense? I should rephrase this to focus on the fact this is about protocol design (the inception stage of a p2p application) not about the deployed stage.

Reviewer: Got it - makes sense.  

Side: lets find/create a place to point to https://docs.rs/libp2p/latest/libp2p/swarm/struct.SwarmBuilder.html#method.max_negotiating_inbound_streams .  Maybe there's a section about transient/negotiating connections and resources and that those should guarded against too.  Go and Rust both have some protections here.

Each stream has some resource cost associated with it. Depending on the
transport and multiplexer, this can be bigger or smaller. Try to avoid having
too many concurrent streams open per peer for your protocol. Instead try to
limit the maximum number of concurrent streams to something reasonable (surely
you don't need >512 streams open at once for a peer?). Multiple concurrent
streams can be useful for logic or to avoid [Head-of-line
blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking), but having too
many streams will offset these benefits.

Using a stream for a short period of time and then closing it is fine. It's
really the number of _concurrent_ streams that you need to be careful of.

> ## Limit the number of connections your application needs

Reviewer: Given connections happen before streams in an application's lifecycle, maybe move this above the section above?

Like streams, each connection has a resource cost associated with it. A
connection will usually represent a peer and a set of protocols with each their
own resource usage. So limiting connections can have a leveraged effect on your
resource usage. 

> In go-libp2p the number of active connections is managed by the

Reviewer: From reading through this doc I don't think it's clear for a user on when to use the connmgr or the resource manager for go-libp2p.

[`connmgr`](https://pkg.go.dev/github.com/libp2p/go-libp2p@v0.21.0/p2p/net/connmgr#BasicConnMgr.Protect).
`ConnManager` will trim connections when you hit the high watermark number of
connections. You can protect certain connections with the `.Protect` method.

> In rust-libp2p handlers should implement

Reviewer: For Rust, I think we should draw more from https://www.notion.so/pl-strflt/rust-libp2p-2a62a76b60c54bd69aea2aa3760d6efe .  At the minimum we should get a link to https://docs.rs/libp2p/latest/libp2p/swarm/struct.ConnectionLimits.html .

[`connection_keep_alive`](https://docs.rs/libp2p/0.46.1/libp2p/swarm/trait.ConnectionHandler.html#tymethod.connection_keep_alive)
to define when a connection can be closed.

## Reduce blast radius

If you can split up your libp2p application into multiple separate processes you
can increase the resiliency of your overall system. For example your node may
have to help achieve consensus and respond to user queries. By splitting this up
into two processes you now rely on the OS’s guarantee that the user query
process won’t take down the consensus process.

## Fail2ban

If you can log when a peer is misbehaving or is malicious, you can then hook up
those logs to fail2ban and have fail2ban manage your firewall to automatically
block misbehaving nodes. go-libp2p includes some builtin support for this
usecase. More details below.

> ## Leverage the resoure manager to limit resource (go-libp2p only)

Reviewer: ```suggestion
## Leverage the resoure manager to limit resource usage (go-libp2p only)
```

Also update TOC

go-libp2p includes a powerful [resource
manager](https://github.com/libp2p/go-libp2p-resource-manager) that keeps track 
of resources used for each protocol, peer, connection, and more. You can use it
within your protocol implementation to make sure you don't allocate more than
some predetermined amount of memory per connection. It's basically a resource
accounting abstraction that you can make use of in your own application.

## Rate limiting incoming connections (go-libp2p only)

Depending on your use case, it can help to limit the number of inbound
connections. You can use go-libp2p's
[ConnectionGater](https://pkg.go.dev/github.com/libp2p/go-libp2p-core/connmgr#ConnectionGater)
and `InterceptAccept` for this. For a concrete example, take a look at how Prysm
implements their (Connection

> Gater)[https://github.com/prysmaticlabs/prysm/blob/63a8690140c00ba6e3e4054cac3f38a5107b7fb2/beacon-chain/p2p/connection_gater.go#L43].

Author: whoops it should be [label](link)

3. Many nodes connecting to a single node such that that node can no
   longer accept new connections from an honest peer. This node is now
   isolated from the honest peers in the network. This is commonly called an
   eclipse attack and is viable if it's either cheap to eclipse this node, or if
   eclipsing this node has a high payoff.

Generally, the effect on our application can range from crashing to stalling to
failing to handle new peers to degraded performance. Ideally we want
our application to at worst suffer a slight performance penalty, but otherwise
stay up and healthy.

In the next section we'll cover some design strategies you should incorporate
into your protocol to make sure your application stays up and healthy.

# Incorporating DOS mitigation from the start

The general strategy is to use the minimum amount of resources as possible and
make sure that there's no untrusted amplification mechanism (e.g. an untrusted
node can force you to do 10x the work it does). A protocol-level reputation
system can help (take a look at [GossipSub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) for inspiration) as well as

> logging misbehaving nodes and actioning those logs separately (see fail2ban
> below).
> 
> Here are some more specific recommendations

Reviewer: ```suggestion
logging misbehaving nodes and actioning those logs separately 
(see [fail2ban](#how-to-automate-blocking-with-fail2ban) below).

Below are some more specific recommendations.
```

## Limit the number of connections your application needs

Each connection has a resource cost associated with it. A connection will
usually represent a peer and a set of protocols with each their own resource
usage. So limiting connections can have a leveraged effect on your resource
usage.

> In go-libp2p the number of active connections is managed by the

Reviewer: A few nitpics here:
1. We use connmgr and ConnManager
2. When we're hyperlinking, I think it's good to remove the ticks so it's clear that it's a hyperlink.  (see screenshot of rendering)
3. We do "ConnManager" no space and "Resource Manager" with space.
4. Maybe the first time we talk about Resource Manager here we make it a hyperlink?

<img width="827" alt="image" src="https://user-images.githubusercontent.com/85411/184450457-8eb69cb9-012f-4ed8-982d-d1f1bda8dbde.png">

Author: > We do "ConnManager" no space and "Resource Manager" with space.

In go-libp2p-core they're called `ConnManager` and `ResourceManager`. Using `Conn Manager` feels weird and so does `ResourceManager` although happy to make one change or the other if you feel strongly.

> When we're hyperlinking, I think it's good to remove the ticks so it's clear that it's a hyperlink. (see screenshot of rendering)

I think our template should support the fixed width + hyperlink. I'll see if I can fix it.

Author: > I think our template should support the fixed width + hyperlink. I'll see if I can fix it.

Fixed this by adding a `text-decoration: underline` css property.

<img width="786" alt="Screen Shot 2022-08-15 at 4 25 20 PM" src="https://user-images.githubusercontent.com/594035/184740948-2049b6e9-258d-4cdc-8735-58d54ba28f65.png">

## Fail2ban

If you can log when a peer is misbehaving or is malicious, you can then hook up
those logs to fail2ban and have fail2ban manage your firewall to automatically
block misbehaving nodes. go-libp2p includes some built-in support for this
use case. More details below.

## Leverage the resource manager to limit resource usage (go-libp2p only)

go-libp2p includes a powerful [resource
manager](https://github.com/libp2p/go-libp2p-resource-manager) that keeps track 
of resources used for each protocol, peer, connection, and more. You can use it
within your protocol implementation to make sure you don't allocate more than
some predetermined amount of memory per connection. It's basically a resource
accounting abstraction that you can make use of in your own application.

## Rate limiting incoming connections (go-libp2p only)

Depending on your use case, it can help to limit the number of inbound
connections. You can use go-libp2p's
[ConnectionGater](https://pkg.go.dev/github.com/libp2p/go-libp2p-core/connmgr#ConnectionGater)
and `InterceptAccept` for this. For a concrete example, take a look at how Prysm
implements their (Connection

> Gater)[https://github.com/prysmaticlabs/prysm/blob/63a8690140c00ba6e3e4054cac3f38a5107b7fb2/beacon-chain/p2p/connection_gater.go#L43].

Author: It's actually a misuse of `(foo)[bar]` vs `[foo](bar)`
